x-pack/plugins/lens/public/editor_frame_service/editor_frame/config_panel/draggable_dimension_button.tsx

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useMemo } from 'react';
+import { DragDrop, DragDropIdentifier, DragContextState } from '../../../drag_drop';
+import { Datasource, VisualizationDimensionGroupConfig, isDraggedOperation } from '../../../types';
+import { LayerDatasourceDropProps } from './types';
+
+const isFromTheSameGroup = (el1: DragDropIdentifier, el2?: DragDropIdentifier) =>
+  el2 && isDraggedOperation(el2) && el1.groupId === el2.groupId && el1.columnId !== el2.columnId;
+
+const isSelf = (el1: DragDropIdentifier, el2?: DragDropIdentifier) =>
+  isDraggedOperation(el2) && el1.columnId === el2.columnId;
+
+export function DraggableDimensionButton({
+  layerId,
+  label,
+  accessorIndex,
+  groupIndex,
+  layerIndex,
+  columnId,
+  group,
+  onDrop,
+  children,
+  dragDropContext,
+  layerDatasourceDropProps,
+  layerDatasource,
+}: {
+  dragDropContext: DragContextState;
+  layerId: string;
+  groupIndex: number;
+  layerIndex: number;
+  onDrop: (droppedItem: DragDropIdentifier, dropTarget: DragDropIdentifier) => void;
+  group: VisualizationDimensionGroupConfig;
+  label: string;
+  children: React.ReactElement;
+  layerDatasource: Datasource<unknown, unknown>;
+  layerDatasourceDropProps: LayerDatasourceDropProps;
+  accessorIndex: number;
+  columnId: string;
+}) {
+  const value = useMemo(() => {
+    return {
+      columnId,
+      groupId: group.groupId,
+      layerId,
+      id: columnId,
+    };
+  }, [columnId, group.groupId, layerId]);
+
+  const { dragging } = dragDropContext;
+  const dragType = isSelf(value, dragging)
+    ? 'move'
+    : isDraggedOperation(dragging) && group.groupId === dragging.groupId
+    ? 'reorder'
+    : 'copy';
+
+  const dropType = isDraggedOperation(dragging)
+    ? group.groupId !== dragging.groupId
+      ? 'replace'
+      : 'reorder'
+    : 'add';
+
+  const isCompatibleFromOtherGroup =
+    dragging?.groupId !== group.groupId &&
+    layerDatasource.canHandleDrop({
+      ...layerDatasourceDropProps,
+      columnId,
+      filterOperations: group.filterOperations,
+    });
+
+  const isDroppable = isDraggedOperation(dragging)
+    ? dragType === 'reorder'
+      ? isFromTheSameGroup(value, dragging)
+      : isCompatibleFromOtherGroup
+    : layerDatasource.canHandleDrop({
+        ...layerDatasourceDropProps,
+        columnId,
+        filterOperations: group.filterOperations,
+      });
```

Comment:
nitpick: assigning to these 3 some names it makes a little bit easier to track the different paths in the various conditions flags:

```ts
const isCurrentGroup = group.groupId === dragging.groupId;
const isDragging = isDraggedOperation(dragging);
const layerCanHandleDrop = layerDatasource.canHandleDrop({
      ...layerDatasourceDropProps,
      columnId,
      filterOperations: group.filterOperations,
    })

const dragType = isSelf(value, dragging)
	? 'move' 
    : isDragging && isCurrentGroup
    ? 'reorder'
    : 'copy';

const dropType = isDragging
    ? !isCurrentGroup
      ? 'replace'
      : 'reorder'
    : 'add';

const isCompatibleFromOtherGroup = !isCurrentGroup && layerCanHandleDrop;

const isDroppable = isDragging
    ? dragType === 'reorder'
      ? isFromTheSameGroup(value, dragging)
      : isCompatibleFromOtherGroup
    : layerCanHandleDrop
```